src/uct/sm/mm/mm_xpmem.c

```diff
@@ -0,0 +1,198 @@
+/**
+ * Copyright (c) UT-Battelle, LLC. 2014-2015. ALL RIGHTS RESERVED.
+ * Copyright (C) Mellanox Technologies Ltd. 2001-2015.  ALL RIGHTS RESERVED.
+ * $COPYRIGHT$
+ * $HEADER$
+ */
+
+#include "mm_pd.h"
+#include "mm_iface.h"
+
+#include <ucs/debug/memtrack.h>
+#include <ucs/debug/log.h>
+#include "xpmem.h"
+
+
+static ucs_status_t uct_xpmem_query()
+{
+    int fd, ver;
+
+    fd = open(XPMEM_DEV_PATH, O_RDWR);
+    if (fd < 0) {
+        ucs_debug("Could not open the XPMEM device file at /dev/xpmem: %m. Disabling xpmem resource");
+        return UCS_ERR_UNSUPPORTED;
+    }
+    close(fd);
+
+    ver = xpmem_version();
+    if (ver < 0) {
+        ucs_debug("Failed to query XPMEM version %d, %m", ver);
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    return UCS_OK;
+}
+
+static ucs_status_t uct_xmpem_reg(void *address, size_t size, 
+                                  uct_mm_id_t *mmid_p)
+{
+    xpmem_segid_t segid; /* 64bit ID */
+    const size_t page_size = ucs_get_page_size();
+    void*  addr_aligned   = (void *)ucs_align_down((uintptr_t)address, page_size);
+    off_t  diff = (uintptr_t)address - (uintptr_t)addr_aligned;
+    size_t length_aligned = ucs_align_up(diff + size, page_size);
+
+    ucs_assert_always(address >= addr_aligned);
+
+    segid = xpmem_make(addr_aligned, length_aligned, XPMEM_PERMIT_MODE, (void *)0666);
```

Comment:
This is the mode that they use in all codes : test, mpi, etc. I would prefer not to touch it.